Selective pruning using hashmask

Greg-Griffith opened the pull request:
This PR adds random block pruning. 
This PR adds a new configurable param which when enabled, will set the node to prune based on a randomly generated 64bit number (henceforth hashmask). A second configurable param called threshold (integer between 1 and 100, 100 by default) can be set to lower threshold by which blocks get prune. 
formula is as follows:
given 
normalized_threshold = threshold * MAX_UINT64/100
then
if (blockhash_64LSB XOR hashmask) < normalized_threshold
keep block
else 
prune block


__TODO__
- if both prune and prunewithmask are set, return an error with an explaination
- if previously had prune or prune with mask, return an error when switching to the other

sickpig commented:
@Greg-Griffith need rebase. w/o it travis won't be executed.

Greg-Griffith commented:
this PR is on hold until a decision about XVersion is made since this feature is completely useless without it (or the addition of a different new message type).

sickpig commented:
@Greg-Griffith restarted all timed out jobs. You still need to fix a formatting problem thou.

Greg-Griffith commented:
rebased

ptschip commented:
Since you resolved your network issue maybe you can drop those last two python test commits...IMO it;s valuable to keep the tests in their original form if we can, for one because they will run faster and two, because they can help to trigger real issues that need debugging.

Greg-Griffith commented:
> Since you resolved your network issue maybe you can drop those last two python test commits...IMO it;s valuable to keep the tests in their original form if we can, for one because they will run faster and two, because they can help to trigger real issues that need debugging.

I dropped those commits from the PR. 

My counter argument for this would be it would solve some annoyances with travis instability, but you make a compelling point about triggering real issues that need debugging which I agree is more important.

ptschip commented on qa/rpc-tests/randompruning.py:
```diff
@@ -0,0 +1,89 @@
+        normalized_threshold = threshold * threshold_percent
+
+        kept = []
+        for block in blocks_mined:
```
Could maybe add some comments here as to what this test is doing and what the expected results are.

Greg-Griffith replied:
added.

ptschip commented on src/allowed_args.cpp:
```diff
@@ -310,10 +310,15 @@ static void addGeneralOptions(AllowedArgs &allowedArgs, HelpMessageMode mode)
             strprintf(_("Reduce storage requirements by pruning (deleting) old blocks. This mode is incompatible with "
-                        "-txindex and -rescan. "
+                        "-txindex and -rescan. Incompatible with -prunewithmask"
                         "Warning: Reverting this setting requires re-downloading the entire blockchain. "
```
Nit:  Should it maybe better to say "This mode is incompatible with -txindex, -rescan and -prunewithmask"  .

Greg-Griffith replied:
fixed

ptschip commented on src/allowed_args.cpp:
```diff
@@ -310,10 +310,15 @@ static void addGeneralOptions(AllowedArgs &allowedArgs, HelpMessageMode mode)
+        .addArg("prunewithmask={true,false,0,1}", optionalBool,
+            strprintf(_("Reduce storage requirements by pruning (deleting) old blocks. This mode is incompatible with "
+                        "-txindex and -rescan. Incompatible with -prune "
+                        "Warning: Reverting this setting requires re-downloading the entire blockchain. "
```
Nit, same this here "This mode is incompatible with -txindex, -rescand and -prune" ?

Greg-Griffith committed:
refactor pruning code into prune.h and prune.cpp

blockdb no longer has a pruneDB method, it is its own separate function
in prune.cpp now

Greg-Griffith committed:
make MIN_BLOCKS_TO_KEEP a chainparam (this helps a lot with qa testing)

Greg-Griffith committed:
add eraseblock and eraseundo methods to blockdb abstract for pruning

Greg-Griffith committed:
replace MIN_BLOCKS_TO_KEEP global var with its new chainparams method

Greg-Griffith committed:
make new hashmask pruning keys for xversion

Greg-Griffith committed:
add methods to relay new xupdates to peers

Greg-Griffith committed:
implement random pruning vars, hashmask generation, comparison method

Greg-Griffith committed:
move pruning setup into prune.cpp and add hashmask logic

Greg-Griffith committed:
store new random pruning variable info on disk

Greg-Griffith committed:
make hashmaskthreshold a tweak, implement a validatior

Greg-Griffith committed:
keep track of peer random pruning data, relay our data when we connect

Greg-Griffith committed:
rework some pruning check logic

Greg-Griffith committed:
add random pruning info to getblockchaininfo if applicable to the node

Greg-Griffith committed:
add args and help text for using new -prunewithmask option

Greg-Griffith committed:
add randompruning.py test

ptschip commented:
looks like a small formatting issue.  I'll try to look at this in the next few days.

ptschip commented on src/rpc/blockchain.cpp:
```diff
@@ -1122,15 +1122,20 @@ UniValue getblockchaininfo(const UniValue &params, bool fHelp)
         {
-            READLOCK(cs_mapBlockIndex);
+            CBlockIndex *block = chainActive.Tip();
             while (block && block->pprev && (block->pprev->nStatus & BLOCK_HAVE_DATA))
```
why take out the READLOCK()? We need that for checking nStatus.

Greg-Griffith replied:
good catch, i will add that back in.

ptschip commented on src/uint256.cpp:
```diff
@@ -76,6 +77,18 @@ std::string base_blob<BITS>::ToString() const
 
+template <unsigned int BITS>
+std::string base_blob<BITS>::ToBitString()
+{
```
interesting but should maybe go in a separate PR

ptschip commented:
I completed my first pass, will probably need to go over it again.  I'll do some testing first though.

ptschip commented on qa/rpc-tests/randompruning.py:
```diff
@@ -0,0 +1,101 @@
+                assert_raises_rpc_error(-32603, "Block not available (pruned data)", self.nodes[1].getblock, block)
+            else:
+                kept.append(self.nodes[1].getblock(block)['height'])
+
```
shouldn't we also check that we kept the blocks we expected to?

Greg-Griffith replied:
doesnt this already do that? if we dont raise the assert where we think we should, then it implies that are keeping everything we expect to keep?

ptschip commented on qa/rpc-tests/randompruning.py:
```diff
@@ -0,0 +1,101 @@
+        uint64_t_max = (2**64) - 1
+        # calculate 1% which
+        threshold_percent = uint64_t_max / 100
+        blockchaininfo = self.nodes[1].getblockchaininfo()
```
maybe add a test using prune.hashMaskThreshold=<something> and verify that it works.

ptschip replied:
Also add a test where you change prune.hashMaskThreshold to a lower value and then check whether it's actually pruning the blocks correctly.  I seem to think from testing that this actually isn't working but I'm not entirely sure.

Greg-Griffith committed:
disallow using both prune and prunewithmask at the same time

Greg-Griffith committed:
add prunethreshold as an arg

Greg-Griffith committed:
add a low prune threshold to the random pruning py test

Greg-Griffith committed:
add missing database compaction call when pruning data from blockdb

Greg-Griffith committed:
add 10% pruning threshold to random pruning python test

Greg-Griffith committed:
make the comparison between two uint256 instead of uint256 < uint64_t

Greg-Griffith committed:
add missing readlock needed for nStatus

Greg-Griffith committed:
validate hashMaskThreshold changes by default

Greg-Griffith committed:
add hashmask threshold changes to python test

Greg-Griffith committed:
add threshold adjustment checks to pruning python test

Greg-Griffith committed:
extend pruning test, mine 500 more blocks after restart at 5% threshold

Greg-Griffith committed:
fix issue where hashmaskthreshold wouldnt change to proper value

Greg-Griffith committed:
fix issue where threshold was not persistant across restarts

Greg-Griffith committed:
formatting fixes

Greg-Griffith committed:
rebase fixes

gandrewstone commented on qa/rpc-tests/randompruning.py:
```diff
@@ -0,0 +1,153 @@
+        self.nodes[1].set, "prune.hashMaskThreshold=70")
+        # our threshold should still be 10
+        print("checking to make sure nothing has changed...")
+        assert_equal({'prune.hashMaskThreshold': 10}, self.nodes[1].get("prune.hashMaskThreshold"))
```
Having only 100 divisions seems too few, especially to add a feature where we automatically reduce the threshold periodically to keep the blockchain size < a configured size in GB.  This feature is what people really want since they can set it to whatever is appropriate based on their disk size.

We could expose the raw threshold number, we could accept floating point numbers in the CLI, or we could used fixed point -- that is, the threshold is a percent * 1e6  (i.e. 1% -> prune.hashMaskThreshold=1000000) 

EDIT: I just realized that with fine gradations, the constant GB feature could be implemented as a script...

gandrewstone commented on src/allowed_args.cpp:
```diff
@@ -311,10 +312,19 @@ static void addGeneralOptions(AllowedArgs &allowedArgs, HelpMessageMode mode)
                     MIN_DISK_SPACE_FOR_BLOCK_FILES / 1024 / 1024))
+        .addArg("prunewithmask={true,false,0,1}", optionalBool,
+            strprintf(_("Reduce storage requirements by pruning (deleting) old blocks. This mode is incompatible with "
+                        "-txindex, -rescan, and -prune "
```
What happens if these flags are both used?  We should make sure no crash.

Greg-Griffith replied:
i have reworked SetupPruning to address this concern. All pruning mode checks are now done in this function instead of during various different parts of init

gandrewstone commented on src/allowed_args.cpp:
```diff
@@ -311,10 +312,19 @@ static void addGeneralOptions(AllowedArgs &allowedArgs, HelpMessageMode mode)
+        .addArg("prunethreshold=<n>", requiredInt,
+            strprintf(_("A normalized threshold to determine how"
+                        "many blocks to keep when prunning with hashmask (default: %u)"),
+                    DEFAULT_THRESHOLD_PERCENT))
```
If you have this as a tweak, you don't need to .addArg.  All tweaks are automatically added as args.

Greg-Griffith replied:
removed this entry from args list.

gandrewstone commented on src/arith_uint256.cpp:
```diff
@@ -179,6 +180,17 @@ std::string base_uint<BITS>::ToString() const
 }
 
+template <unsigned int BITS>
+std::string base_uint<BITS>::ToBitString() const
```
Add a unit test for this API with known test vectors

Greg-Griffith replied:
removed from PR as it is unused. will put it in its own PR as ptschip suggested.

gandrewstone commented on src/arith_uint256.cpp:
```diff
@@ -179,6 +180,17 @@ std::string base_uint<BITS>::ToString() const
+    std::string bits = "";
+    for (int i = 0; i < WIDTH; i++)
+    {
+        bits = bits + std::bitset<WIDTH>(pn[i]).to_string();
```
how about bits.append(std::bitset<WIDTH>(pn[i]).to_string()) to make this a little more efficient?  I wish we could hand bitset a bigger array and avoid this whole for loop.

Greg-Griffith replied:
see above

gandrewstone commented on src/arith_uint256.h:
```diff
@@ -240,6 +240,7 @@ class base_uint
     void SetHex(const char *psz);
     void SetHex(const std::string &str);
     std::string ToString() const;
+    std::string ToBitString() const;
```
add doc comment

Greg-Griffith replied:
see above

gandrewstone commented on src/blockstorage/blockleveldb.h:
```diff
@@ -123,6 +128,7 @@ class CBlockLevelDB : public CDatabaseAbstract
         pwrapperblock->Sync();
         pwrapperundo->Sync();
     }
+    bool EraseBlock(const std::string &key);
```
add doc comments to all new apis

Greg-Griffith replied:
these apis are documented in the abstract class where i defined the virtual functions

gandrewstone commented on src/blockstorage/blockstorage.cpp:
```diff
@@ -796,15 +747,7 @@ bool FindBlockPos(CValidationState &state,
-        if (CheckDiskSpace(nAddSize))
-        {
-            nDBUsedSpace += nAddSize;
-            if (fPruneMode && nDBUsedSpace >= nPruneTarget)
```
are you removing the old pruning method?  Its seems like you are not because the flag is still in addargs but a lot of that code is being removed...

Greg-Griffith replied:
i did not remove the old pruning method. the code was moved/refactored into prune.cpp to make managing the pruning code easier overall mostly by reducing the amount of public api for pruning code. it is down to only a handful of functions.

gandrewstone commented on src/blockstorage/prune.cpp:
```diff
@@ -0,0 +1,418 @@
+}
+
+/** Get important pruning bits from a block hash and compare their value to our pruning threshold*/
+bool hashMaskCompare(uint256 _blockHash)
```
Let's name this better, esp since its not static.  We want to communicate that its about pruning and what is the sense of the returned boolean.  How about:

isBlockPrunable()

Note that if you use this name, it would also swap the return -- that is true would mean prune, and false means keep.

Greg-Griffith replied:
it isnt declared as a static function, but it is only used in this one file. I will change the name and remove it from the header.

gandrewstone commented on src/blockstorage/prune.cpp:
```diff
@@ -0,0 +1,418 @@
+        }
+        if (!fPruneWithMask && nDBUsedSpace < nPruneTarget)
+        {
+            break;
```
this is implements the old style pruning?

Greg-Griffith replied:
in a nutshell, yes. space pruning hasnt been changed, just refactored.

gandrewstone commented on src/blockstorage/prune.cpp:
```diff
@@ -0,0 +1,418 @@
+            pindexOldest = chainActive.Next(pindexOldest);
+            continue;
+        }
+        unsigned int blockSize = pindexOldest->nDataPos;
```
Add comment here:
// We've chosen to prune this block

Greg-Griffith replied:
added

gandrewstone commented on src/blockstorage/prune.cpp:
```diff
@@ -0,0 +1,418 @@
+        if (!fPruneWithMask && nDBUsedSpace < nPruneTarget)
+        {
+            break;
+        }
```
with information like nDBUsedSpace already existing, I feel that it would be quite easy to allow the hashMaskThreshold tweak to take either a % or a disk size.  The hard part would be parsing it but perhaps we could require "%" or "G" to distinguish.  

Now that we have the desired size, if nDBUsedSpace is < desired, we do nothing.  (perhaps in a later PR we have an RPC call that will fill)
If nDBUsedSpace is > desired-fudgefactor (the fudge factor is so that we don't bump over the limit next block), then we figure out how much to decrease hashMaskThreshold by.  First figure out how much we want nDBUsedSpace to fall by:  (desired-fudgefactor)/nDBUsedSpace.  This is a fraction from 0-1.  Now we eliminate (on average) that many blocks:

hashMaskThreshold = hashMaskThreshold*((desired-fudgefactor)/nDBUsedSpace).

That should keep automatically reducing the hashMaskThreshold to keep nDBUsedSpace < desired.

gandrewstone commented:
We need a "pruning.md" file in docs that basically describes the approach.  

And most importantly explains how to set it up from an end-user perspective.  In other words, what options to use together in the bitcoin.conf file, and what can be modified via tweaks.

Please include the different pruning options (the old style vs the new style).

Greg-Griffith committed:
rename hashMaskCompare to IsBlockHashBelowPruneThreshold for clarity

Greg-Griffith committed:
add comment clarifying that this block has been chosen for pruning

Greg-Griffith committed:
move randompruning.py to normal test script list

Greg-Griffith committed:
remove repeat prunethreshold arg, its a tweak. fix null pointer issue

Greg-Griffith committed:
rework SetupPruning to include all startup pruning mode checks

gandrewstone commented:
After sync, I configured the threshold for 25%.  The CLI refused to work for hours with the message "pruning blockstore...".  Now I'm at 120% CPU (constantly) and the CLI is quite slow.  But neither of these should be the case since pruning just reduces what's available.  So it should be implementable as a background thread that grabs cs_main when it needs to prune a block, deletes the block, then gives cs_main back, and then does a short delay.  

We should not let pruning interrupt normal processing, since all its doing is reducing the disk space.

Greg-Griffith committed:
keep track of what the last pruned block was when pruning in db mode

Greg-Griffith committed:
move the actual db pruning and condensing to its own thread
